### ISSUE TITLE
PUSH queries get blocked when the Vert.x event-loop handling the request is running another push query

### ISSUE THREAD

**Issue by spena** [confluentinc/ksql#7358]
**Describe the bug**
The issue is with the number of PUSH queries that can be executed by KSQL. This issue was found by a user who said some KSQL queries were stuck and not responding. 

The problem is when running several PUSH queries, sometimes some queries do not respond. Consequent queries may or may not respond. Sometimes, the user could run 4 queries in a row, the 5th gets stuck, but the 6th and 7th queries could  be run without issues. Killing one query may or may not affect the 5th query. Sometimes, you kill 1 query, and the 5th is still stuck, but you kill another query, and the 5th query starts running and streaming data. This behavior is not consistent.

I found out the issue is related to how Vert.x handles the PUSH queries requests. Vert.x has two different types of threads, event loop threads and worker threads.

Event loop threads are configured by the `ksql.verticle.instances` configuration (default: 2 * CPU cores). An event loop is a thread that handles a request on the KSQL endpoints. Event loops do not block a request. An event loop thread quickly handles the request and respond to the user. If the request needs to do some heavy work, such as PUSH queries which stream data until terminated, then the event loop puts these requests into a queue, so they are handled by the worker threads.

```
public static final String VERTICLE_INSTANCES = KSQL_CONFIG_PREFIX + "verticle.instances";
  public static final int DEFAULT_VERTICLE_INSTANCES =
      2 * Runtime.getRuntime().availableProcessors();
  public static final String VERTICLE_INSTANCES_DOC =
      "The number of server verticle instances to start per listener. Usually you want at least "
          + "many instances as there are cores you want to use, as each instance is single "
          + "threaded.";
```

Worker threads are configured by the `ksql.worker.pool.size` configuration (default: 100). Worker threads can handle code that may block. So the reason of having 100 works by default. Worker threads are not in sync with event loop threads. An event loop thread puts a request into a queue and returns a response to the client. A worker thread later picks a request from the queue and executes it for as long as it needs to (PUSH queries for instance).

```
public static final String WORKER_POOL_SIZE = KSQL_CONFIG_PREFIX + "worker.pool.size";
  public static final String WORKER_POOL_DOC =
      "Max number of worker threads for executing blocking code";
  public static final int DEFAULT_WORKER_POOL_SIZE = 100;
```

You may think that 100 workers would handle 100 push queries. And that 8 event loop threads (2 x 4 cores) would handle only the requests and return. But there's the problem. Each event loop thread has a TaskQueue, which keeps the requests in order of execution. Workers are shared across event loops. So a worker can take an item from the queue of any event loop thread. However, if a worker is already running an item from an event loop, say event-loop-1, and a new PUSH query is handled by the event-loop-1, then his push query is put into the event-loop-1 TaskQueue, but no workers will execute the new item because one is already running it. Thus causing the PUSH query to block. Later on, another PUSH query is handled by a different event loop with no workers running, then this query will be executed. This causes the confusion with users about some queries running and some not responding.

Solution seems easy to fix. It just needs some discussion and analysis by the team. KSQL uses the `executeBlocking` call here https://github.com/confluentinc/ksql/blob/master/ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/OldApiUtils.java#L135. This calls the method that runs the blocking code sequentially, thus blocking other consequent requests.

To fix it, the `executeBlocking` code must be called with the `false` value to run requests in parallel.
```
<T> void executeBlocking(Handler<Promise<T>> blockingCodeHandler,
                         boolean ordered,
                         Handler<AsyncResult<T>> resultHandler)
...
ordered - if true then if executeBlocking is called several times on the same context, the executions for that context will be executed serially, not in parallel. if false then they will be no ordering guarantees
```

KSQL has other parts that call `executeBlocking` with serial execution too. These need to be revised to check if we can run them in parallel.

**To Reproduce**
1. Start KSQL with the following properties:
```
ksql.worker.pool.size=2
ksql.verticle.instances=2
```

2. Then start two CLI instances, and run one PUSH query per instance. Make sure the queries are returning some data to see the behavior. At this point, both queries are handled by the 2 workers. 

3. Contrl+C the 2nd queries and run it again. This query will not respond. Now only 1 query is running, but the other is blocked.

The reason of the block is that Vert.x chooses the event loop to handle a new request using a round-robin algorithm. So, the 3rd push request will be handled by the 1st event loop, which has a worker running the 1st query. 

**Expected behavior**

The number of blocking code should be based on the `ksql.workers.pool.size`, not the `ksql.verticle.instances`.  If I set 100, then I should run 100 queries.

**Comment by agavra** [comment]
This is an excellent analysis - thanks @spena for digging down to the root of this! Do you think this might be related to https://github.com/confluentinc/ksql/issues/7172 as well?

**Comment by spena** [comment]
@agavra Yes. It is the same behavior, so it is definitely related to this.

**Comment by vvcephei** [comment]
Ok, I've got a fix here: https://github.com/confluentinc/ksql/pull/7420

**Comment by vvcephei** [comment]
This should be fixed in 0.18 via https://github.com/confluentinc/ksql/pull/7420
